### PR TITLE
fix: more accurate chunk start and end offsets

### LINKED ISSUE #86: YouTube chunking timestamps are not quite right
The way YouTube chunking is done means that the timestamps are often only *approximately* correct.

In particular, when we [keep the end of a chunk as the start of the next chunk](https://github.com/FullFact/health-misinfo-shared/blob/840e2a939f82b65e1519ed1644436098239b1c8f/src/health_misinfo_shared/youtube_api.py#L122-L123), we don’t rewind the start offset at all. This means it’s possible that an extracted claim might not have come from text that’s between the start and end timestamp. It also means that when we jump to the timestamp in the original audio, the text and the audio don’t quite match up which makes it a bit difficult for the user to anchor themselves.

It’s much easier to navigate if the timestamps are *exactly* correct, which I think it’s possible to do if we keep track of sentences alongside their original timestamps.

### PR DESCRIPTION
**Pull request by andylolz**
Fixes #86.

One problem with keeping track of the running sentence text is: when we discard everything but the last 500 characters (in order to form overlapping blocks) we lose timestamp offset information.

Here, I’ve changed this approach, to ensure we keep timestamp offset information with the sentence text.

One thing that is potentially weird here is: the final chunk will not have an end offset, because we have no way of knowing that (the end of the final chunk is the end of the video. So if we knew the length of the video, we could set it to that.) I’ve set it to `None` here, but we’ll need to amend other code to deal with this.

-----------------

## Pull request checklist

- [x] I have linked my PR to an issue
- [x] I’ve used [conventional commits](https://github.com/FullFact/automation-docs/wiki/Conventional-commits)
- [x] My branch is up-to-date with `main`
- [ ] Where appropriate, I have added or updated tests
- [ ] Where appropriate, I have [updated documentation](https://github.com/FullFact/automation-docs/) to reflect my changes

### CODE REVIEW

File: src/health_misinfo_shared/youtube_api.py [hunk 1]
```diff
@@ -107,28 +107,24 @@ def load_texts(folder) -> list[dict]:
...
+                "end_offset": sentence["start"],
             }
-            # Keep the end of this chunk as the start of the next...
-            current_chunk_text = current_chunk_text[-500:]
-            # ...but remove the first (probably incomplete) word
-            current_chunk_text[current_chunk_text.index(" ") :].strip()
-            current_chunk_start_offset = sentence["start"]
-    current_chunk_end_offset = sentence["start"]
+            while chunk_length(current_chunk) > 500:
+                current_chunk = current_chunk[1:]
```
Comment (andylolz) [inline src/health_misinfo_shared/youtube_api.py:122]: We discard sentences until the chunk has fewer than 500 characters.

File: src/health_misinfo_shared/youtube_api.py [hunk 2]
```diff
@@ -107,28 +107,24 @@ def load_texts(folder) -> list[dict]:
...
-    current_chunk_start_offset = 0.0
-    current_chunk_end_offset = 0.0
+    def chunk_length(chunk: list[dict]) -> int:
+        return sum((len(c["sentence_text"]) for c in chunk))
+
+    current_chunk = []
     for sentence in transcript:
-        current_chunk_text += sentence["sentence_text"] + " "
-        if len(current_chunk_text) > 1500:
-            current_chunk_end_offset = sentence["start"]
```
Comment (andylolz) [inline src/health_misinfo_shared/youtube_api.py:116]: The sentence has already been appended here, so this end offset is off by one sentence.

File: src/health_misinfo_shared/youtube_api.py [hunk 3]
```diff
@@ -107,28 +107,32 @@ def load_texts(folder) -> list[dict]:
...
-                "end_offset": current_chunk_end_offset,
+                "text": " ".join((c["sentence_text"] for c in current_chunk)),
+                "start_offset": current_chunk[0]["start"],
+                "end_offset": sentence["start"],
             }
-            # Keep the end of this chunk as the start of the next...
-            current_chunk_text = current_chunk_text[-500:]
-            # ...but remove the first (probably incomplete) word
-            current_chunk_text[current_chunk_text.index(" ") :].strip()
-            current_chunk_start_offset = sentence["start"]
```
Comment (andylolz) [inline src/health_misinfo_shared/youtube_api.py:126]: We keep the last 500 characters of text, but this doesn’t necessarily quite correspond to the start offset.

**Review by dcorney — APPROVED** [review]
Looks good - neater code and less wrong!

**Pull request merged**